rfcs/2021-11-03-9862-ingest-apm-stats-along-traces-in-dd-agent-source.md

```diff
@@ -0,0 +1,292 @@
+bytes errorSummary = 11; // ddsketch summary of error spans latencies encoded in protobuf
+bool synthetics = 12; // set to true on spans generated by synthetics traffic
+uint64 topLevelHits = 13; // count of top level spans aggregated in the groupedstats
+```
+
+So it is basically a group of various metrics that can be represented in Vector (Sketches are support since this
+[PR][sketch-pr]). In the proto definition sketches are stored as unstructured bytes slices, but those
+[fields][call-to-toproto] are filled with a [protobuf encoded ddsketch][ddsktech-proto]. Given that sketches in Vector
+are also [heavily based on ddsketch][vector-sketch], APM stats sketches can be converted to/from the Vector internal
+representation without incuring too much accuracy losses, but this would require significant work to implement those
+conversion.
+
+This opens two major very different paths for APM stats in Vector:
+
+* Either the APM stats are emitted as a log, each [`ClientGroupedStats`][client-grouped-stats-proto] would be mapped to
+  one log event.
+* Or they are emitted as metrics each value from every [`ClientGroupedStats`][client-grouped-stats-proto] would be
+  emitted as a metric with all upper level information store as tags, but as each metric event flows independently from
+  other ones this would imply adding a significant re-aggregation logic in the traces sink.
+
+Those to approach can be mixed up together if we introduce one of the following ability to bundle multiple metrics into
+a single event:
+
+* Allow multiple metric samples into a single metric event
+* Allow log events to embed metrics, it would mean adding a `Metric` type to the `Value` enum
+
+This also raises the question of having a second (assuming that the `datadog-agent` sources accepts Datadog Agent
+metrics - [RFC][dd-agent-metric-rfc]), unrelated, metric stream coming out of the `datadog-agent` source. Obviously
+event ingested representing APM stats will have to be routed along with traces, and most often they will follow a
+different path that other plain metrics/logs received from a core agent. Thus it is suggested to re-arrange the
+`datadog-agent` source, many option are available (Additional details on what exactly are "Datadog Agents" can be
+found in the trace support [RFC][trace-support-pr] and may provide relevant context for undermentionned points):
+
+* Keeping a single `datadog-agent` source:
+  * and add a settings to switch between agent kind: `agent: <TYPE>` where `<TYPE>`
+  could be  `core` (would support metrics & logs - we could optionnally add `logs` & `metrics` to only allow logs or
+  metrics, along with `core` that would allow both), `trace` and could be extended to support `process`, `security` and
+  so on
+  * or introduce the ability for a source to have multiple output for the `datadog-agent` source, e.g. <SRC_ID>.metrics, <SRC_ID>.logs, <SRC_ID>.traces, <SRC_ID>.apm_stats, etc.
```

Comment:
This is my preference. When a user defines a `datadog_agent` source it accepts _all_ data and outputs over the various channels. I like the `apm_stats` channel.

Reply:
Likewise. I prefer the multiple outputs.